api/app/Console/Commands/CreateDigitalCareersPoolsCandidates.php

```diff
@@ -0,0 +1,219 @@
+                // a match is when a salary int is MORE than the lower limit AND LESS than the upper limit
+                foreach ($salaryArray as $salaryPair) {
+                    if ($salaryPair[0] <= $minSalaryInt && $salaryPair[1] >= $minSalaryInt) {
+                        return true;
+                    }
+                    if ($salaryPair[0] <= $maxSalaryInt && $salaryPair[1] >= $maxSalaryInt) {
+                        return true;
+                    }
+                }
+            }
+            return false;
+        }
+
+        /**
+         * check whether a candidate can match a pool stream
+         * @return boolean
+        */
+        function candidateMatchesStream(array $userWithAssets, string $stream): bool {
+            global $assetArrayMap, $assetKeyToStreamMap;
+
+            $assetsArray = $userWithAssets['cmo_assets']; // grab array of assets models from the passed in collection
+            // convert array of user assets models to array of asset id's
+            $assetsIdArray = [];
+            foreach ($assetsArray as $asset) {
+                array_push($assetsIdArray, $asset['pivot']['cmo_asset_id']);
+            }
+
+           // having collected an array of asset ids, can iterate through and check for matches using the global variables
+            foreach ($assetsIdArray as $assetId) {
+                $assetKey = $assetArrayMap[$assetId];
+                // to avoid errors where the map is missing a key
+                if (array_key_exists($assetKey, $assetKeyToStreamMap)){
+                    $correspondingStream = $assetKeyToStreamMap[$assetKey];
+                    if ($correspondingStream == $stream) {
+                        return true;
+                    }
+                }
+            }
+            return false;
+        }
+
+        // main part, loop through all the candidates, and loop through all classification/streams, calling above functions
+        // create a new pool candidate instance when both functions return TRUE
+        foreach ($candidateCollection as $candidate) {
+            // organize data related to each candidate first
+            $userId = $candidate->user_id;
+            $currentStatus = $candidate->pool_candidate_status;
+
+            // for classification matching grab the user model to pass in as a whole to act on
+            $userModelForCandidate = User::where('id', $userId)->sole();
+
+            // for stream matching, collect user with assets and then pass in
+            $UserForCandidateWithAssetsCollection = User::where('id', $userId)
+                                                            ->get()
+                                                            ->load(['cmoAssets'])
+                                                            ->toArray();
+            $userWithAssets = $UserForCandidateWithAssetsCollection[0]; // get() returns desired collection shape but as an array of one item
```

Comment:
You're getting the same user twice, here. With the suggestions mentioned above, we can pass $userModelForCandidate to both `candidateMatchesClassification` and `candidateMatchesStream`. If we want to make this slightly more efficient, we can [eager load](https://stackoverflow.com/questions/26005994/laravel-with-method-versus-load-method) the cmoAssets when we get the user initially.
```suggestion
            // We eager load expectedClassifications and cmoAssets because we know we will use them to match Classifications and Streams.
            $userModelForCandidate = User::where('id', $userId)->with('expectedClassifications')->with('cmoAssets')->sole();
```